sway/ipc-server.c

```diff
@@ -843,42 +851,43 @@ void ipc_client_handle_command(struct ipc_client *client) {
 		json_object *json = json_object_new_object();
 		json_object_object_add(json, "config", json_object_new_string(config->current_config));
 		const char *json_string = json_object_to_json_string(json);
-		client_valid =
-			ipc_send_reply(client, json_string, (uint32_t)strlen(json_string));
+		client_valid = ipc_send_reply(client, payload_type, json_string,
+			(uint32_t)strlen(json_string));
 		json_object_put(json); // free
 		goto exit_cleanup;
-    }
+	}
 
 	case IPC_SYNC:
 	{
 		// It was decided sway will not support this, just return success:false
 		const char msg[] = "{\"success\": false}";
-		ipc_send_reply(client, msg, strlen(msg));
+		ipc_send_reply(client, payload_type, msg, strlen(msg));
 		goto exit_cleanup;
 	}
 
 	default:
-		sway_log(SWAY_INFO, "Unknown IPC command type %i", client->current_command);
+		sway_log(SWAY_INFO, "Unknown IPC command type %x", payload_type);
 		goto exit_cleanup;
 	}
 
 exit_cleanup:
 	if (client_valid) {
-		client->payload_length = 0;
+		client->pending_length = 0;
```

Comment:
What do you think about changing `ipc_client_handle_command` to return a success boolean and having this reset in `ipc_client_handle_readable` after the `ipc_client_handle_command` calls? That would make it so the pending properties are only touched in `ipc_client_handle_readable` (aside from the initialization in `ipc_handle_connection` anyway).

Reply:
I was trying to be conservative in the number of changes since this is a critical bug for my utilities, and I didn't want to refactor and potentially introduce bugs. That change does look fine, though. I'm not sure I understand the reason for the code, though.

Reply:
If `ipc_send_reply` fails, it will disconnect and free `client`. Once that happens, `client` should not be used anymore

Reply:
In that case, let's move the assignment to 0 to the start of the function after asserting that client is not null, since there's no reason to wait. We've used the payload_length already, so it should be set to 0 afterward. Technically, we should set `pending_type` to an empty value as well.

Reply:
Could also store it in a local temp variable in `ipc_client_handle_readable`, reset it before the call, and pass the local temp variable

Reply:
And with that, we can drop `client_valid` too, since it's not used.

Reply:
@RedSoxFan yup, considered that too. If I think of `handle_command` like a flush, then I don't think that setting `pending_length` to 0 there is bad practice. At worst, it should be redundant.

Reply:
And for a final cleanup, casting `strlen` to `uint32_t` is inconsistent, which indicates to me that it isn't strictly required (unless it's some platform dependent fix I'm not aware of)

Reply:
I'm still not a huge fan of `client->pending_length` being altered in `handle_command` after the rest of the changes this PR makes. This PR removes all other accesses of it in the function and passes the payload length so that it does not use the property. Having `client->pending_length` modified in the function just seems to go against the rest of the PR

Reply:
@RedSoxFan changed